src/TodoApp.tsx

```diff
@@ -0,0 +1,196 @@
+import React, {
+  useCallback, useEffect, useMemo, useState,
+} from 'react';
+import { useParams } from 'react-router-dom';
+import { Todo } from './Todo';
+import { TodoList } from './TodoList';
+import { TodoFilter } from './TodoFilter';
+import { Filters } from './Filters';
+
+export const TodoApp = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [title, setTitle] = useState('');
+  const [filtered, setFiltered] = useState<Filters>(Filters.All);
+  const { filter } = useParams();
+
+  useEffect(() => {
+    let filterParam;
+
+    switch (filter) {
+      case 'active': filterParam = Filters.Active;
+        break;
+      case 'completed': filterParam = Filters.Completed;
+        break;
+      default: filterParam = Filters.All;
+        break;
+    }
+
+    setFiltered(filterParam);
+  }, [filtered, filter]);
+
+  const filteredTodos = useMemo(() => {
+    let newTodos = todos;
+
+    switch (filtered) {
+      case 'Active':
+        newTodos = newTodos.filter(todo => !todo.completed);
+        break;
+      case 'Completed':
+        newTodos = newTodos.filter(todo => todo.completed);
+        break;
+      case 'All':
+        newTodos = todos;
+        break;
+      default: throw new Error('wrong filters');
+    }
+
+    return newTodos;
+  }, [filtered, todos, filter]);
+
+  const handleCreateTodo = useCallback((event: React.FormEvent) => {
+    event.preventDefault();
+
+    const newTodo = {
+      id: +new Date(),
+      title,
+      completed: false,
+    };
+
+    setTodos(prevState => [...prevState, newTodo]);
+    setTitle('');
+  }, [title]);
+
+  const handleToggle = useCallback((id: number) => {
+    const toggleId = todos.filter(todo => todo.id === id)[0];
+    const changeTodo = {
+      id,
+      title: toggleId.title,
+      completed: !toggleId.completed,
+    };
+
+    setTodos(prevState => [...prevState
+      .filter(todo => todo.id !== id), changeTodo].sort((a, b) => a.id - b.id));
+  }, [todos, filteredTodos]);
+
+  const handleAllToggle = useCallback(() => {
+    const newTodos = todos.find(todo => !todo.completed)
+      ? todos.map(todo => {
+        const toggleTodo = {
+          id: todo.id,
+          title: todo.title,
+          completed: true,
+        };
+
+        return toggleTodo;
+      })
+      : todos.map(todo => {
+        const toggleTodo = {
+          id: todo.id,
+          title: todo.title,
+          completed: false,
```

Comment:
since you use 
`        const toggleTodo = {
          id: todo.id,
          title: todo.title,
          completed: true,
        };`
twice, maybe lets create a function that handles those two cases?